internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,271 @@
+package syslog
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	maxRequestEntries          = 100
+	dialTimeout                = 10 * time.Second
+	canonicalPrivEnterpriseNum = 28978
+)
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	facilityUserLevelMessage = 1
+	severityNotice           = 5
+)
+
+type ClientOptions struct {
+	MaxRequestEntries int
+	TargetName        string
+	Location          string
+	Hostname          string
+	SDID              string
+	DialTimeout       time.Duration
+}
+
+type entryWithService struct {
+	service   string
+	Timestamp string
+	Message   string
+}
+
+type Client struct {
+	options *ClientOptions
+
+	// To store log entries, keep a buffer of size 2*MaxRequestEntries with a
+	// sliding window "entries" of size MaxRequestEntries.
+	buffer  []entryWithService
+	entries []entryWithService
+
+	// Store the custom labels(syslog's structured-data) for each service
+	labels map[string]string
+
+	// connection info
+	conn     net.Conn
+	location *url.URL
+	closed   bool
+	sendBuf  bytes.Buffer
+}
+
+// priorityVal calculates the syslog Priority value (PRIVAL) from the given
+// Facility and Severity values. See RFC 5424, sec 6.2.1 for details.
+func priorityVal(facility, severity int) int {
+	return facility*8 + severity
+}
+
+func fillDefaultOptions(options *ClientOptions) {
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = maxRequestEntries
+	}
+	if options.DialTimeout == 0 {
+		options.DialTimeout = dialTimeout
+	}
+}
+
+// NewClient creates a syslog client.
+func NewClient(options *ClientOptions) (*Client, error) {
+
```

Comment:
Style nit: let's avoid a blank line here (for consistency with other code and typical Go style).